File: tests/omnia/data/cache/SimpleIntCacherTest.java
package omnia.data.cache;

import static com.google.common.truth.Truth.assertThat;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import java.util.function.IntSupplier;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

@RunWith(JUnit4.class)
public class SimpleIntCacherTest {
  @Test
  public void value_didReturnSuppliedValue() {
    int testValue = 132;
    CachedInt testSubject = new SimpleIntCacher(() -> testValue);

    assertThat(testSubject.value()).isEqualTo(testValue);
  }

  @Test
  public void value_twice_didReturnSuppliedValue() {
    int testValue = 132;
    CachedInt testSubject = new SimpleIntCacher(() -> testValue);

    testSubject.value();

    assertThat(testSubject.value()).isEqualTo(testValue);
  }

  @Test
  public void new_didNotInvokeSupplier() {
    IntSupplier supplier = setUpMockSupplier();
    new SimpleIntCacher(supplier);

    verify(supplier, never()).getAsInt();
  }

  @Test
  public void invalidate_didNotSupplier() {
    IntSupplier supplier = setUpMockSupplier();
    CachedInt testSubject = new SimpleIntCacher(supplier);

    testSubject.invalidate();

    verify(supplier, never()).getAsInt();
  }

  @Test
  public void value_didInvokeSupplier() {
    IntSupplier supplier = setUpMockSupplier();
    CachedInt testSubject = new SimpleIntCacher(supplier);

    testSubject.value();

    verify(supplier).getAsInt();
  }

  @Test
  public void value_didInvokeSupplierOnce() {
    IntSupplier supplier = setUpMockSupplier();
    CachedInt testSubject = new SimpleIntCacher(supplier);

    testSubject.value();
    testSubject.value();

    verify(supplier, times(1)).getAsInt();
  }

  @Test
  public void value_thenInvalidate_thenValue_didInvokeSupplierTwice() {
    IntSupplier supplier = setUpMockSupplier();
    CachedInt testSubject = new SimpleIntCacher(supplier);

    testSubject.value();
    testSubject.invalidate();
    testSubject.value();

    verify(supplier, times(2)).getAsInt();
  }

  @Test
  public void new_withNullSupplier_didThrowException() {
    assertThrows(NullPointerException.class, () -> new SimpleIntCacher(null));
  }

  private static IntSupplier setUpMockSupplier() {
    IntSupplier supplier = mock(IntSupplier.class);
    when(supplier.getAsInt()).thenReturn(132);
    return supplier;
  }
}

File: tests/omnia/data/cache/SimpleIntCacherTest.kt
package omnia.data.cache

import com.google.common.truth.Truth.assertThat
import java.util.function.IntSupplier
import kotlin.test.Test

import org.mockito.Mockito

class SimpleIntCacherTest {

  @Test
  fun value_didReturnSuppliedValue() {
    val testValue = 132
    val testSubject: CachedInt = SimpleIntCacher { testValue }
    assertThat(testSubject.value()).isEqualTo(testValue)
  }

  @Test
  fun value_twice_didReturnSuppliedValue() {
    val testValue = 132
    val testSubject: CachedInt = SimpleIntCacher { testValue }
    testSubject.value()
    assertThat(testSubject.value()).isEqualTo(testValue)
  }

  @Test
  fun new_didNotInvokeSupplier() {
    val supplier = setUpMockSupplier()
    SimpleIntCacher(supplier)
    Mockito.verify(supplier, Mockito.never()).asInt
  }

  @Test
  fun invalidate_didNotSupplier() {
    val supplier = setUpMockSupplier()
    val testSubject: CachedInt = SimpleIntCacher(supplier)
    testSubject.invalidate()
    Mockito.verify(supplier, Mockito.never()).asInt
  }

  @Test
  fun value_didInvokeSupplier() {
    val supplier = setUpMockSupplier()
    val testSubject: CachedInt = SimpleIntCacher(supplier)
    testSubject.value()
    Mockito.verify(supplier).asInt
  }

  @Test
  fun value_didInvokeSupplierOnce() {
    val supplier = setUpMockSupplier()
    val testSubject: CachedInt = SimpleIntCacher(supplier)
    testSubject.value()
    testSubject.value()
    Mockito.verify(supplier, Mockito.times(1)).asInt
  }

  @Test
  fun value_thenInvalidate_thenValue_didInvokeSupplierTwice() {
    val supplier = setUpMockSupplier()
    val testSubject: CachedInt = SimpleIntCacher(supplier)
    testSubject.value()
    testSubject.invalidate()
    testSubject.value()
    Mockito.verify(supplier, Mockito.times(2)).asInt
  }

  companion object {

    private fun setUpMockSupplier(): IntSupplier {
      val supplier = Mockito.mock(IntSupplier::class.java)
      Mockito.`when`(supplier.asInt).thenReturn(132)
      return supplier
    }
  }
}